File: ios/Classes/Methods/UserMethods.swift
import Flutter
import TwilioConversationsClient

class UserMethods: NSObject, TWCONUserApi {
    let TAG = "UserMethods"
    
    /// setFriendlyName
    func setFriendlyNameIdentity(_ identity: String?, friendlyName: String?, completion: @escaping (FlutterError?) -> Void) {
        debug("setFriendlyName => identity: \(String(describing: identity))")
        guard let client = SwiftTwilioConversationsPlugin.instance?.client else {
            return completion(
                FlutterError(
                    code: "ERROR",
                    message: "Client has not been initialized.",
                    details: nil))
        }

        guard let identity = identity else {
            return completion(
                FlutterError(
                    code: "MISSING_PARAMS",
                    message: "Missing 'identity' parameter",
                    details: nil))
        }

        guard let friendlyName = friendlyName else {
            return completion(
                FlutterError(
                    code: "MISSING_PARAMS",
                    message: "Missing 'friendlyName' parameter",
                    details: nil))
        }

        client.subscribedUser(withIdentity: identity) { (result: TCHResult, user: TCHUser?) in
            if result.isSuccessful, let user = user {
                user.setFriendlyName(friendlyName) { (TCHResult) in
                    if result.isSuccessful {
                        self.debug("setFriendlyName => onSuccess")
                        completion(nil)
                    } else {
                        let errorMessage = String(describing: result.error)
                        self.debug("setFriendlyName => onError: \(errorMessage)")
                        completion(
                            FlutterError(
                                code: "ERROR",
                                message: "Error setting friendlyName \(friendlyName) for user \(identity): \(errorMessage)",
                                details: nil))
                    }
                }
            } else {
                let errorMessage = String(describing: result.error)
                self.debug("setFriendlyName => onError: \(errorMessage)")
                completion(
                    FlutterError(
                    code: "ERROR",
                    message: "Error locating user \(identity): \(errorMessage)",
                    details: nil))
            }
        }
    }
    
    /// setAttributes
    func setAttributesIdentity(_ identity: String?, attributes: TWCONAttributesData?, completion: @escaping (FlutterError?) -> Void) {
        debug("setAttributes => identity: \(String(describing: identity))")
        guard let client = SwiftTwilioConversationsPlugin.instance?.client else {
            return completion(
                FlutterError(
                    code: "ERROR",
                    message: "Client has not been initialized.",
                    details: nil))
        }

        guard let identity = identity else {
            return completion(
                FlutterError(
                    code: "MISSING_PARAMS",
                    message: "Missing 'identity' parameter",
                    details: nil))
        }

        guard let attributesData = attributes else {
            return completion(
                FlutterError(
                    code: "MISSING_PARAMS",
                    message: "Missing 'attributes' parameter",
                    details: nil))
        }

        var userAttributes: TCHJsonAttributes? = nil
        do {
            userAttributes = try Mapper.pigeonToAttributes(attributesData)
        } catch LocalizedConversionError.invalidData {
            return completion(
                FlutterError(
                    code: "CONVERSION_ERROR",
                    message: "Could not convert \(attributes?.data) to valid TCHJsonAttributes",
                    details: nil)
            )
        } catch {
            return completion(
                FlutterError(
                    code: "TYPE_ERROR",
                    message: "\(attributes?.type) is not a valid type for TCHJsonAttributes.",
                    details: nil)
            )
        }

        client.subscribedUser(withIdentity: identity) { (result: TCHResult, user: TCHUser?) in
            if result.isSuccessful, let user = user {
                user.setAttributes(userAttributes) { (TCHResult) in
                    if result.isSuccessful {
                        self.debug("setAttributes => onSuccess")
                        completion(nil)
                    } else {
                        let errorMessage = String(describing: result.error)
                        self.debug("setAttributes => onError: \(errorMessage)")
                        completion(
                            FlutterError(
                                code: "ERROR",
                                message: "Error setting attributes \(userAttributes) for user \(identity): \(errorMessage)",
                                details: nil))
                    }
                }
            } else {
                let errorMessage = String(describing: result.error)
                self.debug("setAttributes => onError: \(errorMessage)")
                completion(
                    FlutterError(
                    code: "ERROR",
                    message: "Error locating user \(identity): \(errorMessage)",
                    details: nil))
            }
        }
    }
    
    private func debug(_ msg: String) {
        SwiftTwilioConversationsPlugin.debug("\(TAG)::\(msg)")
    }
}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/UserMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.twilio.conversations.CallbackListener
import com.twilio.conversations.ErrorInfo
import com.twilio.conversations.StatusListener
import com.twilio.conversations.User
import twilio.flutter.twilio_conversations.Api
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin

class UserMethods : Api.UserApi {
    private val TAG = "UserMethods"

    override fun setFriendlyName(
        identity: String,
        friendlyName: String,
        result: Api.Result<Void>
    ) {
        debug("setFriendlyName => identity: $identity")
        val client = TwilioConversationsPlugin.client
            ?: return result.error(RuntimeException("Client is not initialized"))

        client.getAndSubscribeUser(identity, object: CallbackListener<User> {
            override fun onSuccess(user: User) {
                user.setFriendlyName(friendlyName, object: StatusListener {
                    override fun onSuccess() {
                        debug("setFriendlyName => onSuccess")
                        result.success(null)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("setFriendlyName => onError: $errorInfo")
                        result.error(RuntimeException(errorInfo.message))
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                debug("setFriendlyName => onError: $errorInfo")
                result.error(RuntimeException(errorInfo.message))
            }
        })
    }

    override fun setAttributes(
        identity: String,
        attributes: Api.AttributesData,
        result: Api.Result<Void>
    ) {
        debug("setAttributes => identity: $identity")
        val client = TwilioConversationsPlugin.client
            ?: return result.error(RuntimeException("Client is not initialized"))
        val userAttributes = Mapper.pigeonToAttributes(attributes)
            ?: return result.error(RuntimeException("Could not convert $attributes to valid Attributes"))

        client.getAndSubscribeUser(identity, object: CallbackListener<User> {
            override fun onSuccess(user: User) {
                user.setAttributes(userAttributes, object: StatusListener {
                    override fun onSuccess() {
                        debug("setAttributes => onSuccess")
                        result.success(null)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("setAttributes => onError: $errorInfo")
                        result.error(RuntimeException(errorInfo.message))
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                debug("setAttributes => onError: $errorInfo")
                result.error(RuntimeException(errorInfo.message))
            }
        })
    }

    fun debug(message: String) {
        TwilioConversationsPlugin.debug("$TAG::$message")
    }
}